includes/Admin/Cross_Origin_Isolation.php

```diff
@@ -0,0 +1,316 @@
+
+namespace Google\Web_Stories\Admin;
+
+use Google\Web_Stories\User_Preferences;
+use Google\Web_Stories\Story_Post_Type;
+use Google\Web_Stories_Dependencies\AmpProject\Dom\Document;
+use Google\Web_Stories\Experiments;
+use WP_Screen;
+
+/**
+ * Class Cross_Origin_Isolation
+ *
+ * @package Google\Web_Stories
+ */
+class Cross_Origin_Isolation {
+
+	/**
+	 * Experiments instance.
+	 *
+	 * @var Experiments Experiments instance.
+	 */
+	private $experiments;
+
+	/**
+	 * Cross_Origin_Isolation constructor.
+	 *
+	 * @param Experiments $experiments Experiments instance.
+	 */
+	public function __construct( Experiments $experiments ) {
+		$this->experiments = $experiments;
+	}
+	/**
+	 * Init
+	 *
+	 * @return void
+	 */
+	public function init() {
+		if ( ! $this->is_needed() || ! $this->is_edit_screen() ) {
+			return;
+		}
+
+		add_action( 'admin_footer-post.php', [ $this, 'admin_footer' ] );
+		add_action( 'admin_footer-post-new.php', [ $this, 'admin_footer' ] );
+		add_action( 'load-post.php', [ $this, 'admin_header' ] );
+		add_action( 'load-post-new.php', [ $this, 'admin_header' ] );
+		add_filter( 'style_loader_tag', [ $this, 'style_loader_tag' ], 10, 3 );
+		add_filter( 'script_loader_tag', [ $this, 'script_loader_tag' ], 10, 3 );
+		add_filter( 'get_avatar', [ $this, 'get_avatar' ], 10, 6 );
+	}
+
+	/**
+	 * Start output buffer and add headers.
+	 *
+	 * @since 1.6.0
+	 *
+	 * @return void
+	 */
+	public function admin_header() {
+		header( 'Cross-Origin-Opener-Policy: same-origin' );
+		header( 'Cross-Origin-Embedder-Policy: require-corp' );
+
+		ob_start();
+	}
+
+	/**
+	 * Do string replacement and output.
+	 *
+	 * @since 1.6.0
+	 *
+	 * @return void
+	 */
+	public function admin_footer() {
+		$html = (string) ob_get_clean();
+
+		echo $this->replace_in_dom( $html ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+	}
+
+	/**
+	 * Process a html string and add attribute attributes to required tags. 
+	 *
+	 * @since 1.6.0
+	 *
+	 * @param string $html HTML document as string.
+	 *
+	 * @return string
+	 */
+	protected function replace_in_dom( $html ) {
+		$document = Document::fromHtml( $html );
+
+		if ( ! $document ) {
+			return $html;
+		}
+
+		$map = [
+			'link'   => 'href',
+			'img'    => 'src',
+			'iframe' => 'src',
+			'script' => 'src',
+		];
+
+		$processed = [];
+		foreach ( $map as $tag => $attribute ) {
+			$tags = $document->getElementsByTagName( $tag );
+			foreach ( $tags as $node ) {
+				$value = $node->getAttribute( $attribute );
+				if ( ! $value ) {
+					continue;
+				}
+
+				// If already processed tag, attribute and value before, skip.
+				if ( isset( $processed[ $tag ][ $attribute ] ) && in_array( $value, $processed[ $tag ][ $attribute ], true ) ) {
+					continue;
+				}
+
+				// Check to see if tag already has attirbute.
+				$cross_origin = $node->getAttribute( 'crossorigin' );
+				if ( $cross_origin ) {
+					continue;
+				}
+
+				$processed[ $tag ][ $attribute ][] = $value;
```

Comment:
Since `$this->add_attribute()` replaces all attributes regardless of the tag name, there's no need to save the `$tag` in `$processed`.

That means:

1. When looking at `link[href]` elements it will also modify `a` tags with the same `href` value
2. When processing `img[src]`, all `iframe` and `script` tags will be processed as well, making the loop through those irrelevant.

In other words:

Just doing `foreach( ['href', 'src'] as $tag ) { ... }` would have the same effect.

Now I don't like the implications this has on tags like `<a>`

Reply:
Good point. Removed.